kong/init.lua

```diff
@@ -98,9 +99,95 @@ local ffi = require "ffi"
 local cast = ffi.cast
 local voidpp = ffi.typeof("void**")
 
+local plugins_map_version
+local plugins_map_semaphore
+
+local PLUGINS_MAP_CACHE_OPTS = { ttl = 0 }
+
+local configured_plugins
 local loaded_plugins
 local schema_state
 
+local function build_plugins_map(db, version)
+  local map = {}
+
+  for plugin, err in db.plugins:each() do
+    if err then
+      return nil, tostring(err)
+    end
+
+    map[plugin.name] = true
+  end
+
+  if version then
+    plugins_map_version = version
+  end
+
+  configured_plugins = map
+
+  return true
+end
+
+local function plugins_map_wrapper()
+  local version, err = kong.cache:get("plugins_map:version",
+                                      PLUGINS_MAP_CACHE_OPTS,
+                                      utils.uuid)
+  if err then
+    ngx_log(ngx_CRIT, "could not ensure plugins map is up to date: ", err)
+
+    return false
+
+  elseif plugins_map_version ~= version then
+    -- try to acquire the mutex (semaphore)
+
+    local timeout = 60
+    if singletons.configuration.database == "cassandra" then
+      -- cassandra_timeout is defined in ms
+      timeout = singletons.configuration.cassandra_timeout / 1000
+
+    elseif singletons.configuration.database == "postgres" then
+      -- pg_timeout is defined in ms
+      timeout = singletons.configuration.pg_timeout / 1000
+    end
+
+    local ok, err = plugins_map_semaphore:wait(timeout)
+
+    if ok then
+      -- we have the lock but we might not have needed it. check the
+      -- version again and rebuild if necessary
+
+      version, err = kong.cache:get("plugins_map:version",
+                                    PLUGINS_MAP_CACHE_OPTS,
+                                    utils.uuid)
+      if err then
+        ngx_log(ngx_CRIT, "could not ensure plugins map is up to date: ", err)
+
+        plugins_map_semaphore:post(1)
+
+        return false
+
+      elseif plugins_map_version ~= version then
+        -- we have the lock and we need to rebuild the map. go go gadget!
+
+        ngx_log(ngx_DEBUG, "rebuilding plugins map")
+
+        local ok, err = build_plugins_map(kong.db, version)
+        if not ok then
+          ngx_log(ngx_CRIT, "could not rebuild plugins map: ", err)
+        end
+      end
+
+      plugins_map_semaphore:post(1)
+    else
+      ngx_log(ngx_CRIT, "could not acquire plugins map update mutex: ", err)
+
+      return false
+    end
```

Comment:
Would you mind hoisting this branch up below the `:wait()` call? We always try to write failure branches first (which are much shorter), as per our style guide, and the rest of this codebase. Also, this avoids the need for two branches to be written, since instead, we can do:
```lua
local ok, err = ...
if not ok then
  -- ...
  return
end

-- rest goes here, unindented
```